libs/ballot-interpreter-nh/src/interpret/interpret_oval_marks.ts

```diff
@@ -1,17 +1,225 @@
 import {
   binarize,
+  crop,
+  diff,
   getImageChannelCount,
-  simpleRemoveNoise,
+  otsu,
+  outline,
+  PIXEL_BLACK,
+  PIXEL_WHITE,
+  ratio,
+  writeImageData,
 } from '@votingworks/image-utils';
-import { GridLayout } from '@votingworks/types';
-import { assert } from '@votingworks/utils';
-import { matchTemplateImage, scoreTemplateMatch } from '../images';
-import { BallotCardGeometry, InterpretedOvalMark } from '../types';
+import { GridLayout, GridPosition } from '@votingworks/types';
+import { assert, format } from '@votingworks/utils';
+import makeDebug, { Debugger } from 'debug';
+import { BallotCardGeometry, InterpretedOvalMark, Point } from '../types';
 import { loc, makeRect, vec } from '../utils';
 import { InterpretBallotCardLayoutResult } from './interpret_ballot_card_layout';
 
+const DEBUG_NAMESPACE_PREFIX = 'ballot-interpreter-nh:oval-marks:';
+
+function makeGridPositionDebug(gridPosition: GridPosition): Debugger {
+  const label = `${DEBUG_NAMESPACE_PREFIX}${gridPosition.side}-row=${
+    gridPosition.column
+  }-column=${gridPosition.row}-${gridPosition.contestId}-${
+    gridPosition.type === 'option'
+      ? gridPosition.optionId
+      : `write-in-${gridPosition.writeInIndex}`
+  }`;
+
+  return makeDebug(label);
+}
+
+function writeDebugImage(
+  debug: Debugger,
+  imageData: ImageData,
+  name: string
+): void {
+  if (debug.enabled) {
+    const filePath = `./debug-${debug.namespace.slice(
+      DEBUG_NAMESPACE_PREFIX.length + 1
+    )}-${name}.png`;
+    debug(`%s`, filePath);
+    void writeImageData(filePath, imageData);
+  }
+}
+
+/**
+ * Score oval mark at a given location.
+ */
+export function scoreOvalMark(
+  imageData: ImageData,
+  ovalTemplate: ImageData,
+  ovalTopLeftPoint: Point,
+  geometry: BallotCardGeometry,
+  threshold: number,
+  debug: Debugger
+): Omit<InterpretedOvalMark, 'gridPosition'> {
+  const maximumOffset = 7;
+  const outlinedOvalTemplate = outline(binarize(ovalTemplate, threshold));
```

Comment:
instead of a scanned bubble template that we then binarize every time... should we maybe use the rendered graphic of the ballot PDF, cropped accordingly, rather than a scan of the bubble, which will inevitably carry some noise / skew? Not a big deal, just a thought.

Reply:
That's actually part of the point. The scans will have that stuff, so I wanted to have an image that is as close as possible to what we're actually going to see coming from the scanner. I'm not saying this is perfect, but I think it's better than trying to pull something from the PDF.